pkg/auth/dependency/provider/password/password.go

```diff
@@ -0,0 +1,64 @@
+package password
+
+import (
+	"encoding/json"
+
+	"github.com/sirupsen/logrus"
+	"github.com/skygeario/skygear-server/pkg/core/db"
+	"github.com/skygeario/skygear-server/pkg/server/skydb"
+	"golang.org/x/crypto/bcrypt"
+)
+
+type Provider interface {
+	CreateEntry(principalID string, authData interface{}, hashedPassword string) error
+}
+
+type ProviderImpl struct {
+	sqlBuilder  db.SQLBuilder
+	sqlExecutor db.SQLExecutor
+	logger      *logrus.Entry
+}
+
+func NewProvider(builder db.SQLBuilder, executor db.SQLExecutor, logger *logrus.Entry) *ProviderImpl {
+	return &ProviderImpl{
+		sqlBuilder:  builder,
+		sqlExecutor: executor,
+		logger:      logger,
+	}
+}
+
+func (p ProviderImpl) CreateEntry(
+	principalID string,
+	authData interface{},
+	plainPassword string,
+) error {
```

Comment:
Currently user with same auth data can signup multiple time, I think we will further update this to prevent user duplication?

Reply:
I've added a unique constraints to auth_data so it would be unique.

I think this would a lot simpler and robust than the old skygear-server, that creating and deleting unique constraints for individual keys. Though the downside is the unique key won't exist on user record anymore.

Reply:
Also since we don't have database transection yet, duplicated auth data would create zombie `_auth_principal` data.